### PR TITLE
[DataObjects][DataTypes] Clarify and eventually unify enrichFieldDefinition vs. enrichLayoutDefinition

### LINKED ISSUE #6814: [DataObjects][DataTypes] Clarify and eventually unify enrichFieldDefinition vs. enrichLayoutDefinition
for example see 
https://github.com/pimcore/pimcore/blob/45ef287afc65e199353e976c91b14b059c5c1959/models/DataObject/ClassDefinition/Data/Select.php#L412 vs https://github.com/pimcore/pimcore/blob/45ef287afc65e199353e976c91b14b059c5c1959/models/DataObject/ClassDefinition/Data/Select.php#L436

- would it not be enough to apply options from options provider in just one of these methods? 
- should the calculation of options not be the same - one time inheritance is explicitly activated, one time not...

### PR DESCRIPTION
**Pull request by weisswurstkanone**
resolves #6814

### CODE REVIEW

File: models/DataObject/ClassDefinition/DynamicOptionsProvider/SelectionProviderTrait.php [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+        }
+    }
+
+    /**
+     * { @inheritdoc }
+     */
+    public function enrichFieldDefinition(/** array */ $context = []) /** : Data */
+    {
+        $this->doEnrichDefinitionDefinition(null, $context, $this->getName(),
+            'fielddefinition', DataObject\ClassDefinition\Helper\OptionsProviderResolver::MODE_SELECT);
```
Comment (dvesh3) [inline models/DataObject/ClassDefinition/DynamicOptionsProvider/SelectionProviderTrait.php]:
```suggestion
        $this->doEnrichDefinitionDefinition(null, $this->getName(),
            'fielddefinition', DataObject\ClassDefinition\Helper\OptionsProviderResolver::MODE_SELECT, $context);
```

File: models/DataObject/ClassDefinition/DynamicOptionsProvider/SelectionProviderTrait.php [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+        return $this;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function enrichLayoutDefinition(/*?Concrete */ $object, /**  array */ $context = []) // : self
+    {
+        $this->doEnrichDefinitionDefinition($object, $context, $this->getName(),
+            'layout', DataObject\ClassDefinition\Helper\OptionsProviderResolver::MODE_SELECT);
```
Comment (dvesh3) [inline models/DataObject/ClassDefinition/DynamicOptionsProvider/SelectionProviderTrait.php]:
```suggestion
        $this->doEnrichDefinitionDefinition($object, $this->getName(),
            'layout', DataObject\ClassDefinition\Helper\OptionsProviderResolver::MODE_SELECT, $context);
```

File: models/DataObject/ClassDefinition/Data/Select.php [hunk 3]
```diff
@@ -439,16 +440,8 @@ public function setOptionsProviderData($optionsProviderData)
...
-            $this->getOptionsProviderClass(),
-            DataObject\ClassDefinition\Helper\OptionsProviderResolver::MODE_SELECT
-        );
-
-        if ($optionsProvider) {
-            $context['fieldname'] = $this->getName();
-            $options = $optionsProvider->{'getOptions'}($context, $this);
-            $this->setOptions($options);
-        }
+        $this->doEnrichDefinitionDefinition(null, $this->getName(),
```
Comment (dvesh3) [inline models/DataObject/ClassDefinition/Data/Select.php:443]: method already defined in `SelectionProviderTrait`

File: models/DataObject/ClassDefinition/Data/Select.php [hunk 4]
```diff
@@ -458,33 +451,8 @@ public function enrichFieldDefinition(/** array */ $context = []) /** : Data */
...
-            DataObject::setGetInheritedValues($inheritanceEnabled);
-            $this->setOptions($options);
-
-            $defaultValue = $optionsProvider->{'getDefaultValue'}($context, $this);
-            $this->setDefaultValue($defaultValue);
-
-            $hasStaticOptions = $optionsProvider->{'hasStaticOptions'}($context, $this);
-            $this->dynamicOptions = !$hasStaticOptions;
-        }
+        $this->doEnrichDefinitionDefinition($object, $this->getName(),
```
Comment (dvesh3) [inline models/DataObject/ClassDefinition/Data/Select.php:454]: method already defined in `SelectionProviderTrait`

File: models/DataObject/ClassDefinition/DynamicOptionsProvider/SelectionProviderTrait.php [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+trait SelectionProviderTrait
+{
+    /**
+     * @param DataObject\Concrete|null $object
+     * @param array $context
+     * @param string $fieldname
+     * @param string $purpose
+     * @param string $mode
+     */
+    protected function doEnrichDefinitionDefinition(/*?Concrete */ $object, string $fieldname, string $purpose, string $mode, /**  array */ $context = []) {
```
Comment (dvesh3) [inline models/DataObject/ClassDefinition/DynamicOptionsProvider/SelectionProviderTrait.php:33]:
```suggestion
    protected function doEnrichDefinitionDefinition(/*?DataObject\Concrete */ $object, string $fieldname, string $purpose, string $mode, /**  array */ $context = []) {
```

File: models/DataObject/ClassDefinition/Data/FieldDefinitionEnrichmentInterface.php [hunk 6]
```diff
@@ -20,6 +20,11 @@
 interface FieldDefinitionEnrichmentInterface
 {
     /**
+     * If in admin mode this method can be implemented to change the fielddefinition whenever
+     * getFieldDefinition() get called on the data type.
+     * One example purpose is to populate or change dynamic settings like the options for select and multiselect fields.
+     * The context param contains contextual information about the container, the field name, etc ...
```
Comment (dvesh3) [inline models/DataObject/ClassDefinition/Data/FieldDefinitionEnrichmentInterface.php:26]: Can we define some contextual info used by internal calls?

Comment (weisswurstkanone) [reply models/DataObject/ClassDefinition/Data/FieldDefinitionEnrichmentInterface.php:26]: What would be the use case ?

File: models/DataObject/ClassDefinition/DynamicOptionsProvider/SelectionProviderTrait.php [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+ * @internal
+ */
+trait SelectionProviderTrait
+{
+    /**
+     * @param DataObject\Concrete|null $object
+     * @param array $context
+     * @param string $fieldname
+     * @param string $purpose
+     * @param string $mode
```
Comment (dvesh3) [inline models/DataObject/ClassDefinition/DynamicOptionsProvider/SelectionProviderTrait.php]:
```suggestion
     * @param DataObject\Concrete|null $object
     * @param string $fieldname
     * @param string $purpose
     * @param string $mode
     * @param array $context
```

**Pull request merged**